pkg/pubsub/adapter/adapter.go

```diff
@@ -18,253 +18,246 @@ package adapter
-	// of a Broker. We currently set the TransformerURI to be the address of the Broker filter pod.
-	// TODO consider renaming transformer as it is confusing.
-	if a.transformer != nil {
-		transformedCTX, transformedEvent, err := a.transformer.Send(ctx, event)
-		rtctx := cloudevents.HTTPTransportContextFrom(transformedCTX)
+		// If there was a reply, we need to send it to the sink.
+		// We then overwrite the initial event we sent.
+		event, err = binding.ToEvent(ctx, respMsg)
 		if err != nil {
-			logger.Errorf("error transforming cloud event %q", event.ID())
-			a.reporter.ReportEventCount(args, rtctx.StatusCode)
-			return err
+			a.logger.Error("Failed to convert response message to event",
+				zap.Any("response", respMsg), zap.Error(err))
+			msg.Ack()
+			return
 		}
-		if transformedEvent == nil {
-			// This doesn't mean there was an error. E.g., the Broker filter pod might not return a response.
-			// Report the returned Status Code and return.
-			logger.Debugf("cloud event %q was not transformed", event.ID())
-			a.reporter.ReportEventCount(args, rtctx.StatusCode)
-			return nil
-		}
-		// Update the event with the transformed one.
-		event = *transformedEvent
+
 		// Update the tracing information to use the span returned by the transformer.
-		ctx = trace.NewContext(ctx, trace.FromContext(transformedCTX))
+		var span *trace.Span
+		ctx, span = a.startSpan(ctx, event)
+		defer span.End()
+		reply = true
 	}
 
-	// Apply CloudEvent override extensions to the outbound event.
-	for k, v := range a.extensions {
-		event.SetExtension(k, v)
+	// Only if the message is not from a reply, then we should add the override extensions.
+	if !reply {
+		// Apply CloudEvent override extensions to the outbound event.
+		// This code will be mainly executed by Sources.
+		for k, v := range a.extensions {
+			event.SetExtension(k, v)
+		}
 	}
 
-	// Send the event and report the count.
-	rctx, r, err := a.outbound.Send(ctx, event)
-	rtctx := cloudevents.HTTPTransportContextFrom(rctx)
-	a.reporter.ReportEventCount(args, rtctx.StatusCode)
+	resp, err := a.sendMsg(ctx, a.sinkURI, (*binding.EventMessage)(event))
 	if err != nil {
-		return err
-	} else if r != nil {
-		resp.RespondWith(nethttp.StatusOK, r)
+		a.logger.Error("Failed to send message to sink", zap.String("address", a.sinkURI), zap.Error(err))
+		msg.Nack()
+		return
 	}
-	return nil
-}
 
-func (a *Adapter) convert(ctx context.Context, m transport.Message, err error) (*cloudevents.Event, error) {
-	logger := logging.FromContext(ctx)
-	logger.Debug("Converting event from transport.")
+	a.reporter.ReportEventCount(args, resp.StatusCode)
 
-	if msg, ok := m.(*cepubsub.Message); ok {
-		return converters.Convert(ctx, msg, a.SendMode, a.AdapterType)
+	if resp.StatusCode/100 != 2 {
+		a.logger.Error("Event delivery failed", zap.Int("StatusCode", resp.StatusCode))
+		msg.Nack()
+		return
```

Comment:
response body also needs to be closed in this case.

Reply:
done, thanks!